### DISCUSSION TITLE
Why FTXUI doesn't have a listbox?

### DISCUSSION THREAD

**Discussion by exoton** [ArthurSonzogni/FTXUI#593] [General]
Hi. This is a very nice library. Thank you.

I was wondering, why doesn't it have a listbox? You may say "menu", but it's not the same, especially if you need to browse a database table with million of records.

**Comment by ArthurSonzogni** [comment]
If needed, I can make `Menu` more performant, by avoiding rendering entry that are likely not going to fit the screen. This would make it possible to render million of records.

What do you intend exactly by `ListBox`?

**Reply by exoton** [reply]
I mean something like [Android RecyclerView](https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.html) or [Win32 ListView](https://learn.microsoft.com/en-us/windows/win32/controls/list-view-control-reference). You have a scrollable list of items, from which you can select one or multiple items. Item selection doesn't necessarily results in an action (although you can get a notification and act on it). Imaging a dialog box with a list of database records and action buttons `Add`, `Delete`, `Edit`.

**Reply by ArthurSonzogni** [reply]
I think you can start from the `checkbox_in_frame` example:
https://arthursonzogni.github.io/FTXUI/examples/?file=component/checkbox_in_frame
https://github.com/ArthurSonzogni/FTXUI/blob/master/examples/component/checkbox_in_frame.cpp

It won't scale to 1M entry. You probably can't do more than ~4k.
If needed I can submit an optimization to go past this number.

**Reply by exoton** [reply]
Thank you for taking the time to discuss this and for your suggestions. I understand that writing a UI toolkit is a huge undertaking.

Multiple selection is a nice to have option, so `Menu` is probably a better choice for me, but it's still a different type of component. And my use-case is an interface to the database which may grow big.

**Reply by mr-mocap** [reply]
I think I have already written a cursory version of what you may be asking for.
It is just a simple "list widget" that contains a list of strings, one per line, and a couple callbacks for being alerted to when the selected item changes.

I made it as a subcomponent for a directory browser widget that I'm currently working on.

It is by no means complete, but I am willing to flesh it out to make it more generally useful.
Let me know what you think!


[list.hpp](https://github.com/mr-mocap/cli-6502-playground/blob/master/app/ui/components/list.hpp)
[list.cpp](https://github.com/mr-mocap/cli-6502-playground/blob/master/app/ui/components/list.cpp)

P.S.  I think I actually also used the `checkbox_in_frame` as an example.

**Reply by exoton** [reply]
Thank's for letting me know. Sorry, I can't promise a prompt feedback at the moment.
